### DISCUSSION TITLE
ContainerConfig.yaml fragments or ContainerSubConfig.yaml recipes

### DISCUSSION THREAD

**Discussion by psychemedia** [mmh352/ou-container-builder#25] [Ideas]
Allowing the use of user-selected base containers provides one way of quickstarting a build. Another way to support reuse would be to have a collection / library of `ContainerConfig.yaml` fragments or recipes that install particular components.

For example, a couple of modules use PostgreSQL so it may be useful to have a simple collection of setup recipes, including one for installing Postgres and required Python packages to access it etc.

A downside of this is that if you have several recipes with different `apt` or `pip` requirements, then you somehow need to merge them assuming you can only have one `pip` and one `apt` key in the YAML file. A large complex `ContainerConfig.yaml` file may not be overly helpful as a crib if you're trying to figure out how they installed one application if you can't easily tease out it's requirements.

Supporting  things like `SubContainerConfig.postgres.yaml` that could be parsed and merged to create a single `ContainerConfig.yaml` would be a route to having reusable components. This would also work well for off-the-shelf subcontainer files that define an application that might be required by different modules in their container.

Admittedly, there is a risk of then starting to go down the route of (re)creating a whole new package management system, worrying about conflicts etc.

A simpler alternative would be to use create a simple config file as the set of different requirements at each level (`apt`, pip` etc) in a naive way, simplifying what's given into a single config file (doing little more than deduping duplicates etc), and then letting folk figure out for themselves what to do if any version clashes arise directly or via dependencies etc.

The tool would then help the user put pre-existing pieces together, but if they don't fit it's up to the user to fettle them until they do.

**Comment by mmh352** [comment]
The ``ou-container-builder`` comes with the concept of "packs". These are pre-defined sets of settings for frequently repeated patterns. Currently there is only a single one defined (the tutorial-server), but where multiple containers need the same set of base packages, these would then be rolled into a pack, which can be added with a single config line. Packs could cover something like the PostgreSQL system + python packages or the necessary packages to install an R kernel, or anything. Packs are basically a short-cut for sets of ``ContainerConfig`` settings that are merged into the settings provided by the developer.

**Reply by psychemedia** [reply]
Okay, I'll have a go at one of those then for postgres and one for mongo (TM351 components); and maybe one for "datascience" (pandas, scipy etc).

I guess that OpenRefine would be another candidate? (Different ways of achieving the same thing may be interesting to compare in terms of ease of creation, ease of reuse, etc.)

**Reply by mmh352** [reply]
Sounds good. What I would suggest is simply building a ``ContainerConfig`` with both postgres and mongo. Then you know which settings are required and these can then be factored out into (in this case) two "packs". OpenRefine would be another candidate, I agree. The one thing I would say is that it is only worth building a pack if at least two modules share requirements.

**Reply by psychemedia** [reply]
So - packs to crib from... I found https://github.com/mmh352/ou-container-builder/blob/default/src/ou_container_builder/templates/tutorial-server.ini but that looks like it's just a config file to fire up an installed tutorial server?

The simple `ContainerConfig.yaml` is also the approach I was going to take, as per the OpenRefine demo. Being able to merge it with, or load it into a parent `ContainerConfig.yaml` still feels like it'd be a nice approach though. (You'd drop the metadata at the top of the included file (module code/presentation, unless you use that as a check that module or presentation code match or maybe allow wild cards in child to say "let me merged with any presentation etc"), then merge in the other bits, doing `apt` and `pip` as a `set()` and adding additional scripts and apps separately.)

**Reply by mmh352** [reply]
There is nowhere to crib any pack definitions from. This is because the actual way the packs will work has not yet been implemented. The single pack that is currently supported is a hack that needs refactoring when we have a better understanding of what a pack.

That is why I suggested just building a ContainerConfig.yaml that works. Then we look at that and factor out those bits that make sense into one or more packs.

Sure, we can consider hierarchical configuration files at some point in the future, but again that adds significant complexity. When we have an actual scenario where supporting the additional complexity is worth having the inheritance functionality, then we will look at that.

**Reply by psychemedia** [reply]
"Then we look at that and factor out those bits that make sense into one or more packs."

If the "pack" were just an extract from a `ContainerConfig.yaml`, then folk could create either a running instance of it from the `ContainerConfig.yaml` file, or invoke it from a parent `ContainerConfig.yaml` file.

I'm intrigued as to how you implement it and how / why it'd be simpler than `ContainerConfig.yaml` files all the way down...

**Reply by mmh352** [reply]
It is essentially the discussion of whether to support multiple inheritance or mixins. Both solutions have advantages, for the moment the architecture uses mixins in the form of "packs".

**Reply by psychemedia** [reply]
Just struck me out on a walk that there is no need to nest the config settings in an intermediate form, because all they do is drive things like the Dockerfile creation and file copying?

If you have `ContainerConfig.yaml`, `ContainerConfig.1.yaml` `ContainerConfig.openrefine.yaml`etc, and require that `ContainerConfig.yaml` has at least `code` and `presentation` elements, treat `ContainerConfig.yaml` as the first element and then sort the remaining `ContainerConfig.*.yaml` files, you can iterate:

```
# generate dockerfile header
# handle apt, pip script elements etc from ContainerConfig.yaml and add to Dockerfile

for subConfig in ContainerConfig.*.yaml files:
  # ignore code and presentation elements
  # handle apt, pip script elements etc from ContainerConfig.X.yaml and add to Dockerfile

# Add footer
RUN chown -R $USER:$GID $HOME
USER $USER
WORKDIR $HOME
ENTRYPOINT ["tini", "-g", "--"]
CMD ["start.sh"]
```

In the looped subconfig files, I would also allow for optional `title` that would add eg `#### {{ title }} #####` to the Dockerfile, or maybe wrap that part of the Dockerfile in `#### Start: {{ title }} #####`  and `#### End: {{ title }} #####` wrapper comments, and perhaps also a `# {{ description }}` ;-) to comment that part of the Dockerfile, eg for debug education, archiving readability etc. 

Having the subconfigs generate their own layers would create smaller, reusable layers (I can't say for certain, but some students appear to get download issues on images at times (not the same as the issue I had), and I think this occurs more often than not with large layers. Maybe because they can take a lot of unpacking. Or maybe because of network weather as folk sit several rooms away from the home router, downloading large Docker images over the barest sniff of wifi, while the rest of the household stream Netflix and engage in online gaming from closer to the router etc...!)

The full subconfig files would also allow allow for the creation of their own standalone images, which might be useful in docker-compose style arrangements.

**Reply by mmh352** [reply]
That might also be an interesting approach. Certainly worth thinking about.

I agree with the aim of keeping the container size as small as possible. That is one of the reasons for starting with a minimal base. If we see that the resulting containers are much too big, then we can look at seeing how we can make more of the image layers shared between images.

**Reply by psychemedia** [reply]
Re: size, I noted that you don't `apt-get clean && rm -rf /var/lib/apt/lists/*`. Any reason?

Also, I tend to `--no-install-recommends` though that could be a config setting.

**Reply by psychemedia** [reply]
(By the by, I started messing w/ my own bells and whistles branch that will let me try out lots of things you don't like for my own purposes;-) Apt-clean step [here](https://github.com/ouseful-PR/ou-container-builder/blob/12f0abb122a560c8b7b7c87b2d2ac11ec409e5ea/src/ou_container_builder/templates/dockerfile/base.jinja2#L63).

**Reply by mmh352** [reply]
The cleanup step sounds sensible. For concrete things like that an issue or a pull-request would be great.

**Comment by psychemedia** [comment]
Just spotted this new candidate alternative to repo2docker image builder: https://discourse.jupyter.org/t/orbiter-image-builder-for-jupyter/9216

Repo: https://github.com/dorgeln/orbiter

Has a single YAML file for config.

How open are you about sharing `ou-container-builder` approach to wider community with an addition to that thread...

**Reply by scmmmh** [reply]
The recursive structure is interesting.

I'm open to sharing the ``ou-container-builder``, but it is very heavily designed for the OU use-case, so not sure how useful it would be to anyone else.

**Reply by psychemedia** [reply]
Okay... so are you happy for me to share it on the discourse forum as a reply to the orbiter post (it can be useful to help folk compare different new approaches and look for patterns).

I'll also try to do a summary of both approaches, and repo2docker, and compare them,

**Reply by scmmmh** [reply]
Sure. No problem.
